cpp/src/hash/hashing.cu

```diff
@@ -0,0 +1,227 @@
+  }
+
+ private:
+  const size_type mask;
+};
+
+template <bool has_nulls>
+std::vector<std::unique_ptr<experimental::table>>
+hash_partition_table(table_view const& input,
+                     table_view const &table_to_hash,
+                     const size_type num_partitions,
+                     rmm::mr::device_memory_resource* mr,
+                     cudaStream_t stream)
+{
+  auto const num_rows = table_to_hash.num_rows();
+  auto row_partition_numbers = rmm::device_vector<size_type>(num_rows);
+
+  // Make an iterator to compute the hash over each row
+  auto const device_input = table_device_view::create(input, stream);
+  auto const hasher = experimental::row_hasher<MurmurHash3_32, has_nulls>(*device_input);
+  auto const hash_iterator = thrust::make_transform_iterator(
+    thrust::make_counting_iterator(0), hasher);
+
+  // If the number of partitions is a power of two, we can compute the partition 
+  // number of each row more efficiently with bitwise operations
+  if (is_power_two(num_partitions)) {
+    // Compute which partition each row belongs to using bitwise partitioner
+    auto partitioner = bitwise_partitioner<hash_value_type>{num_partitions};
+    thrust::transform(rmm::exec_policy(stream)->on(stream), hash_iterator,
+      hash_iterator + num_rows, row_partition_numbers.begin(), partitioner);
+
+  } else {
+    // Compute which partition each row belongs to using modulo partitioner
+    auto partitioner = modulo_partitioner<hash_value_type>{num_partitions};
+    thrust::transform(rmm::exec_policy(stream)->on(stream), hash_iterator,
+      hash_iterator + num_rows, row_partition_numbers.begin(), partitioner);
+  }
+
+  // Scatter input rows to output partitions given the partition map
+  auto const partition_map = column_view(data_type(INT32), num_rows,
+    row_partition_numbers.data().get());
+  auto output = experimental::detail::scatter_to_tables(input, partition_map, mr, stream);
+
+  // Pad with empty tables if we have less than num_partitions
+  if (output.size() < static_cast<size_t>(num_partitions)) {
+    std::generate_n(std::back_inserter(output), num_partitions - output.size(),
+      [&input]() { return experimental::empty_like(input); });
+  }
+
+  return output;
+}
+
+}  // namespace
+
+namespace detail {
+
+std::vector<std::unique_ptr<experimental::table>>
+hash_partition(table_view const& input,
+               std::vector<size_type> const& columns_to_hash,
+               int num_partitions,
+               rmm::mr::device_memory_resource* mr,
+               cudaStream_t stream)
+{
+  std::vector<std::unique_ptr<experimental::table>> output;
+  auto table_to_hash = input.select(columns_to_hash);
+
+  // Return empty vector if there are no partitions or anything to hash
+  if (num_partitions <= 0 || input.num_rows() == 0 || table_to_hash.num_columns() == 0) {
+    return output;
+  }
+
+  cudf::nvtx::range_push("CUDF_HASH_PARTITION", cudf::nvtx::PARTITION_COLOR);
```

Comment:
Why not start the range at the start of the function?

Reply:
I would need to add corresponding range_pop() calls when input.select() throws an exception and when I return early when the input is empty. If we had an RAII NVTX wrapper that could be automatic! But, I think this is a little cleaner anyway since you would only see the range show up when there's actually work being done.

Reply:
If there's no work being done the range will be so short you won't see it.  I agree RAII NVTX is needed. It would also allow you to remove the early declaration of output which would make @codereport happy.

Reply:
Right! Then I could just do this:
`  if (has_nulls(table_to_hash)) {
    return hash_partition_table<true>(
        input, table_to_hash, num_partitions, mr, stream);
  } else {
    return hash_partition_table<false>(
        input, table_to_hash, num_partitions, mr, stream);
  }`

Reply:
I think @jrhemstad had mentioned that he had an RAII wrapper for NVTX already?

Reply:
> If we had an RAII NVTX wrapper that could be automatic!

Yeah, it's on my backburner:
https://github.com/rapidsai/cudf/issues/2397
https://github.com/jrhemstad/nvtx_wrappers